Misleading error message with malformed turbofishes

I typoed the angle brackets when attempting turbofish.

```rust
fn main() {
    println!("{:?}", vec![1, 2, 3].into_iter().collect::<Vec<usize>>>());
}

```

([Playground](https://play.rust-lang.org/?gist=dd291f4a7c8e184f571f8d00f8dc839c&version=nightly&mode=debug&edition=2015))

I expected to see this happen: To get an error indicating that I had an extra closing
angle bracket after my turbofish, or that I had a mismatched delimiter.

Here is what happened instead:

Errors:

```
   Compiling playground v0.0.1 (/playground)
error: field expressions may not have generic arguments
 --> src/lib.rs:2:55
  |
2 |     println!("{:?}", vec![1, 2, 3].into_iter().collect::<Vec<usize>>>());
  |                                                       ^^^^^^^^^^^^^^

error: aborting due to previous error

error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

This error message refers to the case when you attempt to use turbofish on a struct field
rather than a method or function, like `struct.field::<usize>`. However, this error message
is confusing to users because, if you look at the end of the collect call, there _are_ parentheses,
and this _looks_ like a correctly formed method call, which obscures the real error.

For reference, this is how the compiler responds to the same typo elsewhere.

```rust
fn main() {
    println!("{:?}", Vec::<usize>>::new());
}

```

([Playground](https://play.rust-lang.org/?gist=c194cac89405ed870fa110b94c2a9973&version=nightly&mode=debug&edition=2015))

```
   Compiling playground v0.0.1 (/playground)
error[E0423]: expected value, found struct `Vec`
 --> src/main.rs:2:22
  |
2 |     println!("{:?}", Vec::<usize>>::new());
  |                      ^^^^^^^^^^^^ did you mean `Vec { /* fields */ }`?

error[E0425]: cannot find function `new` in the crate root
 --> src/main.rs:2:37
  |
2 |     println!("{:?}", Vec::<usize>>::new());
  |                                     ^^^ not found in the crate root
```

--- a/src/libsyntax/ast.rs
+++ b/src/libsyntax/ast.rs
@@ -140,6 +140,20 @@ pub enum GenericArgs {
 }
 
 impl GenericArgs {
+    pub fn is_parenthesized(&self) -> bool {
+        match *self {
+            Parenthesized(..) => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_angle_bracketed(&self) -> bool {
+        match *self {
+            AngleBracketed(..) => true,
+            _ => false,
+        }
+    }
+
     pub fn span(&self) -> Span {
         match *self {
             AngleBracketed(ref data) => data.span,

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -2757,6 +2757,8 @@ impl<'a> Parser<'a> {
     // Assuming we have just parsed `.`, continue parsing into an expression.
     fn parse_dot_suffix(&mut self, self_arg: P<Expr>, lo: Span) -> PResult<'a, P<Expr>> {
         let segment = self.parse_path_segment(PathStyle::Expr, true)?;
+        self.check_trailing_angle_brackets(&segment);
+
         Ok(match self.token {
             token::OpenDelim(token::Paren) => {
                 // Method call `expr.f()`
@@ -2782,6 +2784,101 @@ impl<'a> Parser<'a> {
                 self.mk_expr(span, ExprKind::Field(self_arg, segment.ident), ThinVec::new())
             }
         })
+    }
+
+    /// This function checks if there are trailing angle brackets and produces
+    /// a diagnostic to suggest removing them.
+    ///
+    /// ```ignore (diagnostic)
+    /// let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>();
+    ///                                                        ^^ help: remove extra angle brackets
+    /// ```
+    fn check_trailing_angle_brackets(&mut self, segment: &PathSegment) {
+        // This function is intended to be invoked from `parse_dot_suffix` where there are two
+        // cases:
+        //
+        // - A field access (eg. `x.foo`)
+        // - A method call (eg. `x.foo()`)
+        //
+        // This function is called after parsing `.foo` and before parsing any parenthesis (if
+        // present). This includes any angle bracket arguments, such as `.foo::<u32>`.
+
+        // We only care about trailing angle brackets if we previously parsed angle bracket
+        // arguments. This helps stop us incorrectly suggesting that extra angle brackets be
+        // removed in this case:
+        //
+        // `x.foo >> (3)` (where `x.foo` is a `u32` for example)
+        //
+        // This case is particularly tricky as we won't notice it just looking at the tokens -
+        // it will appear the same (in terms of upcoming tokens) as below (since the `::<u32>` will
+        // have already been parsed):
+        //
+        // `x.foo::<u32>>>(3)`
+        let parsed_angle_bracket_args = segment.args
+            .as_ref()
+            .map(|args| args.is_angle_bracketed())
+            .unwrap_or(false);
+
+        debug!(
+            "check_trailing_angle_brackets: parsed_angle_bracket_args={:?}",
+            parsed_angle_bracket_args,
+        );
+        if !parsed_angle_bracket_args {
+            return;
+        }
+
+        // Keep the span at the start so we can highlight the sequence of `>` characters to be
+        // removed.
+        let lo = self.span;
+
+        // We need to look-ahead to see if we have `>` characters without moving the cursor forward
+        // (since we might have the field access case and the characters we're eating are
+        // actual operators and not trailing characters - ie `x.foo >> 3`).
+        let mut position = 0;
+
+        // The first tokens we will encounter are shift right tokens (`>>`) since pairs of `>`
+        // characters will have been grouped together by the tokenizer.
+        let mut number_of_shr = 0;
+        while self.look_ahead(position, |t| *t == token::BinOp(token::BinOpToken::Shr)) {
+            number_of_shr += 1;
+            position += 1;
+        }
+
+        // Afterwards, there will be at most one `>` character remaining (more than one and it'd
+        // have shown up as a `>>`).
+        let encountered_gt = self.look_ahead(position, |t| *t == token::Gt);
+        if encountered_gt {
+            position += 1;
+        }
+
+        // If we didn't find any trailing `>>` characters or a trailing `>`, then we have
+        // nothing to error about.
+        debug!(
+            "check_trailing_angle_brackets: encountered_gt={:?} number_of_shr={:?}",
+            encountered_gt, number_of_shr,
+        );
+        if !encountered_gt && number_of_shr < 1 {
+            return;
+        }
+
+        // Finally, double check that we have a left parenthesis next as otherwise this is the
+        // field case.
+        if self.look_ahead(position, |t| *t == token::OpenDelim(token::Paren)) {
+            // Eat from where we started until the left parenthesis so that parsing can continue
+            // as if we didn't have those extra angle brackets.
+            self.eat_to_tokens(&[&token::OpenDelim(token::Paren)]);
+            let span = lo.until(self.span);
+
+            self.diagnostic()
+                .struct_span_err(span, "unmatched angle bracket")
+                .span_suggestion_with_applicability(
+                    span,
+                    "remove extra angle bracket",
+                    String::new(),
+                    Applicability::MachineApplicable,
+                )
+                .emit();
+        }
     }
 
     fn parse_dot_or_call_expr_with_(&mut self, e0: P<Expr>, lo: Span) -> PResult<'a, P<Expr>> {

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -2869,11 +2869,18 @@ impl<'a> Parser<'a> {
             self.eat_to_tokens(&[&token::OpenDelim(token::Paren)]);
             let span = lo.until(self.span);
 
+            // We needn't check `encountered_gt` to determine if we should pluralize "bracket".
+            // `encountered_gt` can only represent a single `>` character, if `number_of_shr >= 1`
+            // then there is either `>>` or `>>>` - in either case a plural is warranted.
+            let plural = number_of_shr >= 1;
             self.diagnostic()
-                .struct_span_err(span, "unmatched angle bracket")
+                .struct_span_err(
+                    span,
+                    &format!("unmatched angle bracket{}", if plural { "s" } else { "" }),
+                )
                 .span_suggestion_with_applicability(
                     span,
-                    "remove extra angle bracket",
+                    &format!("remove extra angle bracket{}", if plural { "s" } else { "" }),
                     String::new(),
                     Applicability::MachineApplicable,
                 )

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -2149,7 +2149,27 @@ impl<'a> Parser<'a> {
                            enable_warning: bool)
                            -> PResult<'a, ()> {
         loop {
-            segments.push(self.parse_path_segment(style, enable_warning)?);
+            let segment = self.parse_path_segment(style, enable_warning)?;
+            if style == PathStyle::Expr {
+                // In order to check for trailing angle brackets, we must have finished
+                // recursing (`parse_path_segment` can indirectly call this function),
+                // that is, the next token must be the highlighted part of the below example:
+                //
+                // `Foo::<Bar as Baz<T>>::Qux`
+                //                      ^ here
+                //
+                // As opposed to the below highlight (if we had only finished the first
+                // recursion):
+                //
+                // `Foo::<Bar as Baz<T>>::Qux`
+                //                     ^ here
+                //
+                // `PathStyle::Expr` is only provided at the root invocation and never in
+                // `parse_path_segment` to recurse and therefore can be checked to maintain
+                // this invariant.
+                self.check_trailing_angle_brackets(&segment, token::ModSep);
+            }
+            segments.push(segment);
 
             if self.is_import_coupler() || !self.eat(&token::ModSep) {
                 return Ok(());
@@ -2757,7 +2777,7 @@ impl<'a> Parser<'a> {
     // Assuming we have just parsed `.`, continue parsing into an expression.
     fn parse_dot_suffix(&mut self, self_arg: P<Expr>, lo: Span) -> PResult<'a, P<Expr>> {
         let segment = self.parse_path_segment(PathStyle::Expr, true)?;
-        self.check_trailing_angle_brackets(&segment);
+        self.check_trailing_angle_brackets(&segment, token::OpenDelim(token::Paren));
 
         Ok(match self.token {
             token::OpenDelim(token::Paren) => {
@@ -2793,15 +2813,19 @@ impl<'a> Parser<'a> {
     /// let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>();
     ///                                                        ^^ help: remove extra angle brackets
     /// ```
-    fn check_trailing_angle_brackets(&mut self, segment: &PathSegment) {
-        // This function is intended to be invoked from `parse_dot_suffix` where there are two
+    fn check_trailing_angle_brackets(&mut self, segment: &PathSegment, end: token::Token) {
+        // This function is intended to be invoked after parsing a path segment where there are two
         // cases:
         //
-        // - A field access (eg. `x.foo`)
-        // - A method call (eg. `x.foo()`)
+        // 1. A specific token is expected after the path segment.
+        //    eg. `x.foo(`, `x.foo::<u32>(` (parenthesis - method call),
+        //        `Foo::`, or `Foo::<Bar>::` (mod sep - continued path).
+        // 2. No specific token is expected after the path segment.
+        //    eg. `x.foo` (field access)
         //
-        // This function is called after parsing `.foo` and before parsing any parenthesis (if
-        // present). This includes any angle bracket arguments, such as `.foo::<u32>`.
+        // This function is called after parsing `.foo` and before parsing the token `end` (if
+        // present). This includes any angle bracket arguments, such as `.foo::<u32>` or
+        // `Foo::<Bar>`.
 
         // We only care about trailing angle brackets if we previously parsed angle bracket
         // arguments. This helps stop us incorrectly suggesting that extra angle brackets be
@@ -2836,43 +2860,47 @@ impl<'a> Parser<'a> {
         // actual operators and not trailing characters - ie `x.foo >> 3`).
         let mut position = 0;
 
-        // The first tokens we will encounter are shift right tokens (`>>`) since pairs of `>`
-        // characters will have been grouped together by the tokenizer.
+        // We can encounter `>` or `>>` tokens in any order, so we need to keep track of how
+        // many of each (so we can correctly pluralize our error messages) and continue to
+        // advance.
         let mut number_of_shr = 0;
-        while self.look_ahead(position, |t| *t == token::BinOp(token::BinOpToken::Shr)) {
-            number_of_shr += 1;
+        let mut number_of_gt = 0;
+        while self.look_ahead(position, |t| {
+            trace!("check_trailing_angle_brackets: t={:?}", t);
+            if *t == token::BinOp(token::BinOpToken::Shr) {
+                number_of_shr += 1;
+                true
+            } else if *t == token::Gt {
+                number_of_gt += 1;
+                true
+            } else {
+                false
+            }
+        }) {
             position += 1;
         }
 
-        // Afterwards, there will be at most one `>` character remaining (more than one and it'd
-        // have shown up as a `>>`).
-        let encountered_gt = self.look_ahead(position, |t| *t == token::Gt);
-        if encountered_gt {
-            position += 1;
-        }
-
-        // If we didn't find any trailing `>>` characters or a trailing `>`, then we have
-        // nothing to error about.
+        // If we didn't find any trailing `>` characters, then we have nothing to error about.
         debug!(
-            "check_trailing_angle_brackets: encountered_gt={:?} number_of_shr={:?}",
-            encountered_gt, number_of_shr,
+            "check_trailing_angle_brackets: number_of_gt={:?} number_of_shr={:?}",
+            number_of_gt, number_of_shr,
         );
-        if !encountered_gt && number_of_shr < 1 {
+        if number_of_gt < 1 && number_of_shr < 1 {
             return;
         }
 
-        // Finally, double check that we have a left parenthesis next as otherwise this is the
-        // field case.
-        if self.look_ahead(position, |t| *t == token::OpenDelim(token::Paren)) {
-            // Eat from where we started until the left parenthesis so that parsing can continue
+        // Finally, double check that we have our end token as otherwise this is the
+        // second case.
+        if self.look_ahead(position, |t| {
+            trace!("check_trailing_angle_brackets: t={:?}", t);
+            *t == end
+        }) {
+            // Eat from where we started until the end token so that parsing can continue
             // as if we didn't have those extra angle brackets.
-            self.eat_to_tokens(&[&token::OpenDelim(token::Paren)]);
+            self.eat_to_tokens(&[&end]);
             let span = lo.until(self.span);
 
-            // We needn't check `encountered_gt` to determine if we should pluralize "bracket".
-            // `encountered_gt` can only represent a single `>` character, if `number_of_shr >= 1`
-            // then there is either `>>` or `>>>` - in either case a plural is warranted.
-            let plural = number_of_shr >= 1;
+            let plural = number_of_gt > 1 || number_of_shr >= 1;
             self.diagnostic()
                 .struct_span_err(
                     span,

--- a/src/test/ui/issues/issue-54521-1.rs
+++ b/src/test/ui/issues/issue-54521-1.rs
@@ -1,0 +1,16 @@
+// compile-pass
+
+// This test checks that the `remove extra angle brackets` error doesn't happen for some
+// potential edge-cases..
+
+struct X {
+    len: u32,
+}
+
+fn main() {
+    let x = X { len: 3 };
+
+    let _ = x.len > (3);
+
+    let _ = x.len >> (3);
+}

--- a/src/test/ui/issues/issue-54521-2.fixed
+++ b/src/test/ui/issues/issue-54521-2.fixed
@@ -1,0 +1,22 @@
+// run-rustfix
+
+// This test checks that the following error is emitted and the suggestion works:
+//
+// ```
+// let _ = Vec::<usize>>>::new();
+//                     ^^ help: remove extra angle brackets
+// ```
+
+fn main() {
+    let _ = Vec::<usize>::new();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = Vec::<usize>::new();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = Vec::<usize>::new();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = Vec::<usize>::new();
+    //~^ ERROR unmatched angle bracket
+}

--- a/src/test/ui/issues/issue-54521-2.rs
+++ b/src/test/ui/issues/issue-54521-2.rs
@@ -1,0 +1,22 @@
+// run-rustfix
+
+// This test checks that the following error is emitted and the suggestion works:
+//
+// ```
+// let _ = Vec::<usize>>>::new();
+//                     ^^ help: remove extra angle brackets
+// ```
+
+fn main() {
+    let _ = Vec::<usize>>>>>::new();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = Vec::<usize>>>>::new();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = Vec::<usize>>>::new();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = Vec::<usize>>::new();
+    //~^ ERROR unmatched angle bracket
+}

--- a/src/test/ui/issues/issue-54521-2.stderr
+++ b/src/test/ui/issues/issue-54521-2.stderr
@@ -1,0 +1,26 @@
+error: unmatched angle brackets
+  --> $DIR/issue-54521-2.rs:11:25
+   |
+LL |     let _ = Vec::<usize>>>>>::new();
+   |                         ^^^^ help: remove extra angle brackets
+
+error: unmatched angle brackets
+  --> $DIR/issue-54521-2.rs:14:25
+   |
+LL |     let _ = Vec::<usize>>>>::new();
+   |                         ^^^ help: remove extra angle brackets
+
+error: unmatched angle brackets
+  --> $DIR/issue-54521-2.rs:17:25
+   |
+LL |     let _ = Vec::<usize>>>::new();
+   |                         ^^ help: remove extra angle brackets
+
+error: unmatched angle bracket
+  --> $DIR/issue-54521-2.rs:20:25
+   |
+LL |     let _ = Vec::<usize>>::new();
+   |                         ^ help: remove extra angle bracket
+
+error: aborting due to 4 previous errors
+

--- a/src/test/ui/issues/issue-54521.fixed
+++ b/src/test/ui/issues/issue-54521.fixed
@@ -1,0 +1,22 @@
+// run-rustfix
+
+// This test checks that the following error is emitted and the suggestion works:
+//
+// ```
+// let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>();
+//                                                        ^^ help: remove extra angle brackets
+// ```
+
+fn main() {
+    let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>();
+    //~^ ERROR unmatched angle bracket
+}

--- a/src/test/ui/issues/issue-54521.rs
+++ b/src/test/ui/issues/issue-54521.rs
@@ -1,0 +1,22 @@
+// run-rustfix
+
+// This test checks that the following error is emitted and the suggestion works:
+//
+// ```
+// let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>();
+//                                                        ^^ help: remove extra angle brackets
+// ```
+
+fn main() {
+    let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>>>();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>>();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>();
+    //~^ ERROR unmatched angle bracket
+
+    let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>();
+    //~^ ERROR unmatched angle bracket
+}

--- a/src/test/ui/issues/issue-54521.stderr
+++ b/src/test/ui/issues/issue-54521.stderr
@@ -1,0 +1,26 @@
+error: unmatched angle bracket
+  --> $DIR/issue-54521.rs:11:60
+   |
+LL |     let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>>>();
+   |                                                            ^^^^ help: remove extra angle bracket
+
+error: unmatched angle bracket
+  --> $DIR/issue-54521.rs:14:60
+   |
+LL |     let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>>();
+   |                                                            ^^^ help: remove extra angle bracket
+
+error: unmatched angle bracket
+  --> $DIR/issue-54521.rs:17:60
+   |
+LL |     let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>();
+   |                                                            ^^ help: remove extra angle bracket
+
+error: unmatched angle bracket
+  --> $DIR/issue-54521.rs:20:60
+   |
+LL |     let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>();
+   |                                                            ^ help: remove extra angle bracket
+
+error: aborting due to 4 previous errors
+

--- a/src/test/ui/issues/issue-54521.stderr
+++ b/src/test/ui/issues/issue-54521.stderr
@@ -1,20 +1,20 @@
-error: unmatched angle bracket
+error: unmatched angle brackets
   --> $DIR/issue-54521.rs:11:60
    |
 LL |     let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>>>();
-   |                                                            ^^^^ help: remove extra angle bracket
+   |                                                            ^^^^ help: remove extra angle brackets
 
-error: unmatched angle bracket
+error: unmatched angle brackets
   --> $DIR/issue-54521.rs:14:60
    |
 LL |     let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>>();
-   |                                                            ^^^ help: remove extra angle bracket
+   |                                                            ^^^ help: remove extra angle brackets
 
-error: unmatched angle bracket
+error: unmatched angle brackets
   --> $DIR/issue-54521.rs:17:60
    |
 LL |     let _ = vec![1, 2, 3].into_iter().collect::<Vec<usize>>>>();
-   |                                                            ^^ help: remove extra angle bracket
+   |                                                            ^^ help: remove extra angle brackets
 
 error: unmatched angle bracket
   --> $DIR/issue-54521.rs:20:60

